envelope.go

```diff
@@ -67,14 +67,32 @@ func (w *envelopeWriter) Marshal(message any) *Error {
 		}
 		return nil
 	}
-	raw, err := w.codec.Marshal(message)
+
+	buffer := w.bufferPool.Get()
+
+	// Reuse byte buffer if codec supports MarshalAppend.
+	var (
+		raw []byte
+		err error
+	)
+	if c, ok := w.codec.(marshalAppender); ok {
+		raw, err = c.MarshalAppend(buffer.Bytes(), message)
+	} else {
+		raw, err = w.codec.Marshal(message)
+	}
+	if err == nil && len(raw) > 0 && cap(raw) != cap(buffer.Bytes()) {
+		// We've either grown the buffer or allocated an entirely new one. Before
+		// we return, capture the new backing array in the buffer pool.
+		defer w.bufferPool.Put(bytes.NewBuffer(raw))
+	} else {
+		// The original buffer was sufficiently large for the workload, so we can
+		// return it to the pool.
+		defer w.bufferPool.Put(buffer)
```

Comment:
Do we? For best performance, we want the buffers in the pool to be large enough to contain the marshaled bytes without resizing (at least most of the time). We _know_ that the buffer we've pulled from the pool was too small for this message. I think we're better off letting that buffer be GC'ed and putting the new buffer into the pool instead, under the assumption that we need the buffers in the pool to grow to accommodate the application workload.

None of this is perfect, but I think we're probably better off pooling over- rather than under-aggressively. The GC is still free to empty the pool on each cycle if it needs to reclaim the space. (Last I looked, I think it just categorically emptied pools...but I haven't checked in many years.)

Reply:
🤦🏽 @pkwarren pointed out to me that we're calling `buffer.Write` just a few lines below, which allocates the same amount of memory all over again. That makes returning the original buffer to the pool fine, but also means we're eating allocs that we definitely don't need.

Reply:
Calling `bytes.NewBuffer(largerThanBuf)` will cause an allocation like `buffer.Write(largerThanBuf)` but has the chance for reuse. Maybe it's better to split these two methods to keep the old behaviour

```go
	var buffer *bytes.Buffer
	defer w.bufferPool.Put(buffer)

	switch codec := w.codec.(type) {
	case marshalAppender:
		buffer = w.bufferPool.Get()
		raw, err := codec.MarshalAppend(buffer.Bytes(), message)
		if err != nil {
			return errorf(CodeInternal, "marshal append message: %w", err)
		}
		buffer.Write(raw)
	default:
		raw, err := w.codec.Marshal(message)
		if err != nil {
			return errorf(CodeInternal, "marshal message: %w", err)
		}
		// We can't avoid allocating the byte slice, so we may as well reuse it once
		// we're done with it.
		buffer = bytes.NewBuffer(raw)
	}
	envelope := &envelope{Data: buffer}
	return w.Write(envelope)
```

Reply:
Did something similar-ish in the next commit to more clearly separate the two paths. I think I've gotten the best of both worlds: we avoid allocating if `MarshalAppend` didn't grow the slice capacity, and we eat a fixed-size allocation instead of an O(N) reallocation if it did. Performance improved slightly:

```
[main] BenchmarkConnect/unary-12                   8054           1318347 ns/op        14414394 B/op        251 allocs/op
[original PR] BenchmarkConnect/unary-12            10192          1176274 ns/op         6122383 B/op        236 allocs/op
[latest] BenchmarkConnect/unary-12                 10448          1153636 ns/op         6115680 B/op        236 allocs/op
```

Reply:
Looks good to me!